src/main/scala/org/ergoplatform/nodeView/history/storage/modifierprocessors/ToDownloadProcessor.scala

```diff
@@ -36,20 +37,26 @@ trait ToDownloadProcessor extends BasicReaders with ScorexLogging {
     */
   def isHeadersChainSynced: Boolean = pruningProcessor.isHeadersChainSynced
 
-  /** Returns Next `howMany` modifier ids satisfying `filter` condition our node should download
-    * to synchronize full blocks
+  /**
+    * Get modifier ids to download to synchronize full blocks
+    * @param howManyPerType how many ModifierIds per ModifierTypeId to fetch
+    * @param condition filter only ModifierIds that pass this condition
+    * @return next max howManyPerType ModifierIds by ModifierTypeId to download filtered by condition
     */
-  def nextModifiersToDownload(howMany: Int, condition: ModifierId => Boolean): Seq[(ModifierTypeId, ModifierId)] = {
+  def nextModifiersToDownload(howManyPerType: Int, condition: ModifierId => Boolean): Map[ModifierTypeId, Seq[ModifierId]] = {
     @tailrec
-    def continuation(height: Int, acc: Seq[(ModifierTypeId, ModifierId)]): Seq[(ModifierTypeId, ModifierId)] = {
-      if (acc.lengthCompare(howMany) >= 0) {
-        acc.take(howMany)
+    def continuation(height: Int, acc: Map[ModifierTypeId, Seq[ModifierId]]): Map[ModifierTypeId, Seq[ModifierId]] = {
+      // return if at least one of Modifier types reaches howManyPerType limit for modifier ids
+      if (acc.values.exists(_.lengthCompare(howManyPerType) >= 0)) {
+        acc.mapValues(_.take(howManyPerType)).view.force
       } else {
         val headersAtThisHeight = headerIdsAtHeight(height).flatMap(id => typedModifierById[Header](id))
 
         if (headersAtThisHeight.nonEmpty) {
           val toDownload = headersAtThisHeight.flatMap(requiredModifiersForHeader).filter(m => condition(m._2))
-          continuation(height + 1, acc ++ toDownload)
+          // add new modifiers to download to accumulator
+          val newAcc = toDownload.foldLeft(acc) { case (newAcc, (mType, mId)) => newAcc.adjust(mType)(_.fold(Seq(mId))(_ :+ mId)) }
```

Comment:
Looks like using `_ :+ mId` in a loop makes this quadratic i.e. O(toDownload.size ^ 2).
What is the typical size of toDownload?
Consider using some builder here and make both the logic more explicit and performance better. 
It is not functional though, but I would prefer performance, since this is system level code.

Reply:
Ah, sorry, frankly I haven't used `Seq()` anywhere in past 10 years and I thought it was backed by `Vector`, I've always gone for concrete constructor in the past or Builders, but they don't have `size/length` information.

There is just 5-20 elements in it, I guess it would deserve a rewrite to something more efficient ...

Reply:
@kushti @aslesarenko guys let me do this in next iteration, I just changed it to `Vector` from `List` for now, for instance `HeadersProcessor` is full of `Seq()` and `Seq.empty` ... I did not even know about this constructor